config/config.go

```diff
@@ -322,12 +323,20 @@ func (s *QTypeSet) UnmarshalYAML(unmarshal func(interface{}) error) error {
 var validDomain = regexp.MustCompile(
 	`^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9])\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\-]*[A-Za-z0-9])$`)
 
-// ParseUpstream creates new Upstream from passed string in format [net]:host[:port][/path]
+// ParseUpstream creates new Upstream from passed string in format [net]:host[:port][/path][#commonname]
 func ParseUpstream(upstream string) (Upstream, error) {
 	var path string
 
 	var port uint16
 
+	var commonName *string
+
+	cn, upstream := extractCN(upstream)
+
+	if cn != "" {
+		commonName = &cn
+	}
```

Comment:
If you go with both of my above solutions, this can become:
```suggestion
	upstream, commonName := extractCN(upstream)
```